0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,146 @@
+- **Forwarding, Not Delivery** - In contrast to the first version of ILP, ILPv4 does not have the destination amount in the packet and connectors must only apply their local rate to forward packets. Senders may use higher-level protocols to indicate to the receiver the minimum amount they should accept for each packet, but this is no longer built into the core protocol. Connectors have only one simple behavior and do not need to maintain up-to-date price information on all possible destintations.
+- **Quoting is an Application Concern** - Connectors are only responsible for forwarding ILP packets and do not need to implement a [separate protocol for quoting](../0008-interledger-quoting-protocol/0008-interledger-quoting-protocol.md). Applications can use test packets to determine the exchange rate of a particular path.
+
+## Flow
+
+Before the protocol starts, the sender and receiver use a higher-level protocol to agree on the destination account and packet condition.
+
+1. Sender constructs an ILP Prepare packet with the receiver's destination account, the agreed-upon condition, and an amount and expiry of their choice. The sender may include additional data, the formatting of which is determined by the higher-level protocol.
+2. Sender sends the ILP Prepare packet to a connector over some communication channel, such as a Websocket. 
+3. Connector checks whether the sender has sufficient bandwidth (balance or credit) to send the amount specified in the packet. If so, the connector reduces the sender's bandwidth by the value of the packet. If not, the connector returns an ILP Reject packet.
+4. Connector uses the destination ILP address from the packet and its local routing tables to determine which next hop to forward the packet to.
+5. Connector modifies the packet amount and expiry to apply their exchange rate and shorten the expiry time.
+6. Connector forwards the packet to the next hop, which may be another connector. All subsequent connectors go through steps 3-6 until the packet reaches the receiver.
+7. Receiver checks the ILP Prepare packet, according to whatever is stipulated by the higher-level protocol (such as checking whether the amount received is above some minimum specified by the sender). 
+8. Receiver returns an ILP Fulfill packet, containing the preimage of the condition in the ILP Prepare packet. The receiver may include additional data, the formatting of which is determined by the higher-level protocol.
+9. Connector checks that the receiver has returned the fulfillment before the expiry in the ILP Prepare packet. If so, the connector returns the same ILP Fulfill packet to the previous connector and credits the receiver's account with the amount in the ILP Prepare packet. If the Prepare has expired, the connector returns an ILP Reject packet to the previous connector. Each connector repeats this step until the Fulfill (or Reject) packet reaches the sender.
```

Comment:
>If the Prepare has expired, the connector returns an ILP Reject packet to the previous connector.
Why can't the previous connector just determine that the Prepare expired when they don't get a Fulfill packet before the expiration?